### PR TITLE
Optimize BytesReferenceStreamInput Number Reads

### LINKED ISSUE #70800: Query performance degrade after 7.12 in big aggregation case
We encounter query performance degrade after we upgrade to 7.12. The performance is almost 2 times slower than old version.
Below is the flame graph after in 7.12
image

You can see `BytesReferenceStreamInput.maybeNextSlice` may be the performance bottleneck now and it is introduced in https://github.com/elastic/elasticsearch/pull/65921.

The query is 
```
{
  "query": {
    "bool": {
      "filter": [
        {
          "terms": {
            "xxxx_id": [
              xxxxx
              ]
          }
        },
        {
          "range": {
            "local_stat_time_hour": {
              "lt": "2020-12-30 00:00:00",
              "gte": "2020-10-01 00:00:00",
              "time_zone": "+00:00"
            }
          }
        }
      ]
    }
  },
  "aggs": {
    "stat_time_hour": {
      "date_histogram": {
        "min_doc_count": 1,
        "format": "yyyy-MM-dd HH:mm:ss",
        "interval": "hour",
        "extended_bounds": {
          "max": "2020-12-30 00:00:00",
          "min": "2020-10-01 00:00:00"
        },
        "field": "local_stat_time_hour"
      },
      "aggs": {
        "campaign_id": {
          "terms": {
            "field": "zzz_id",
            "size": 50000,
            "min_doc_count": 1
          },
          "aggs": {
            "yyyy": {
              "sum": {
                "field": "yyyy"
              }
            },
     ............... total 92 sum aggs
          }
        }
      }
    }
  },
  "size": 0
}
```

In 7.6.2, it takes 15 seconds to run the query but in 7.12, it takes 30 seconds. The response size is 600MB.
I guess if we can construct a query which produce large response size and many sum aggs, then we can reproduce the issue.

### PR DESCRIPTION
**Pull request by original-brownbear**
`readByte` was used for all primitive reads which via the expensive `maybeNextSlice`
which did not inline well caused slowness when reading large aggregate messages.

This commit makes these reads do bulk operations on the underlying `byte[]` where possible
instead. Also, it extracts cold path from `maybeNextSlice` so that we get a fast bounds check
branch instead of the loop with a large body for the hot case.

closes #70800

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java [hunk 1]
```diff
@@ -40,26 +41,64 @@
 
     @Override
     public byte readByte() throws IOException {
-        if (offset() >= bytesReference.length()) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:43]: no need for double checking here, the slice lengths must be correct (it's an assumption we make in other places so the `maybeNextSlice` is all we need

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java [hunk 2]
```diff
@@ -40,26 +41,64 @@
...
+    }
+
     protected int offset() {
         return sliceStartOffset + sliceIndex;
     }
 
     private void maybeNextSlice() throws IOException {
-        while (sliceIndex == slice.length) {
-            sliceStartOffset += sliceIndex;
+        if (sliceIndex == slice.length) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:86]: Its worth leaving a comment here that we're being super intentional about moving the slow stuff to another method so the `if` statement is inlined.

**Review by nik9000 — COMMENTED** [review]
This looks right to me. I think it's worth a jmh benchmark to prove to ourselves that we're really helping.

It isn't a "for now" thing, but I'd love to be able to assert that the jvm *can* inline methods we think are important to inline. We go to a lot of trouble to juggle things so the JVM can inline them but we don't have any way to test whether or not we've broken them. Some kind of test would be nice. But that ain't a "for now" thing.

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:86]: ++ added one.

**Review by nik9000 — APPROVED** [review]

**Comment by original-brownbear** [comment]
Thanks Nik!

**Pull request merged**